12-offer-encoding.md

```diff
@@ -0,0 +1,1057 @@
+2. types:
+    1. type: 2 (`chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 4 (`offer_id`)
+    2. data:
+        * [`sha256`:`offer_id`]
+    1. type: 8 (`amount`)
+    2. data:
+        * [`tu64`:`msat`]
+    1. type: 12 (`features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 32 (`quantity`)
+    2. data:
+        * [`tu64`:`quantity`]
+    1. type: 36 (`recurrence_counter`)
+    2. data:
+        * [`tu32`:`counter`]
+    1. type: 68 (`recurrence_start`)
+    2. data:
+        * [`tu32`:`period_offset`]
+    1. type: 38 (`payer_key`)
+    2. data:
+        * [`point32`:`key`]
+    1. type: 39 (`payer_note`)
+    2. data:
+        * [`...*utf8`:`note`]
+    1. type: 50 (`payer_info`)
+    2. data:
+        * [`...*byte`:`blob`]
+    1. type: 56 (`replace_invoice`)
+    2. data:
+        * [`sha256`:`payment_hash`]
+    1. type: 240 (`payer_signature`)
+    2. data:
+        * [`bip340sig`:`sig`]
+
+## Requirements for Invoice Requests
+
+The writer of an invoice_request:
+  - MUST set `payer_key` to a transient public key.
+  - MUST remember the secret key corresponding to `payer_key`.
+  - MUST set `offer_id` to the Merkle root of the offer as described in [Signature Calculation](#signature-calculation).
+  - MUST NOT set or imply any `chain_hash` not set or implied by the offer.
+  - MUST set `payer_signature` `sig` as detailed in [Signature Calculation](#signature-calculation) using the `payer_key`.
+  - if the offer had a `quantity_min` or `quantity_max` field:
+    - MUST set `quantity`
+    - MUST set it within that (inclusive) range.
+  - otherwise:
+    - MUST NOT set `quantity`
+  - if the offer did not specify `amount`:
+    - MUST specify `amount`.`msat` in multiples of the minimum lightning-payable unit
+      (e.g. milli-satoshis for bitcoin) for the first `chains` entry.
+  - otherwise:
+    - MAY omit `amount`.
+    - if it sets `amount`:
+      - MUST specify `amount`.`msat` as greater or equal to amount expected by the offer
+        (before any proportional period amount).
+  - if the sender has a previous unpaid invoice (for the same offer) which it wants to cancel:
+    - MUST set `payer_key` to the same as the previous invoice.
+    - MUST set `replace_invoice` to the `payment_hash` or the previous invoice.
+  - if the offer contained `recurrence`:
+    - for the initial request:
+      - MUST use a unique `payer_key`.
+      - MUST set `recurrence_counter` `counter` to 0.
+    - for any successive requests:
+      - MUST use the same `payer_key` as the initial request.
+      - MUST set `recurrence_counter` `counter` to one greater than the highest-paid invoice.
+    - if the offer contained `recurrence_base` with `start_any_period` non-zero:
+      - MUST include `recurrence_start`
+      - MUST set `period_offset` to the period the sender wants for the initial request
+      - MUST set `period_offset` to the same value on all following requests.
```

Comment:
Does this mean that the request is for period number `recurrence_start.period_offset + recurrence_counter.counter`?